### PR TITLE
Add Transformer Block in Search Space

### PR DESCRIPTION
**Pull request by achaar**
### Which issue(s) does this Pull Request fix?

resolves #000

### Details of the Pull Request

To add a Transformer Block in the Search Space. A multi-head self-attention block is required for the Transformer, also a Token and Position embedding layer might be needed.

### CODE REVIEW

**Comment by achaar** [comment]
This is not the final commit. Just wanted to know if the direction of implementing this is okay or not. I have to add comments and unit test case. Also, for this to work for text especially, we also need Token and Position embedding as input. Any suggestions in this regard are welcome.

File: autokeras/tuners/task_specific.py [hunk 1]
```diff
@@ -50,6 +50,24 @@
     'text_block_1/dense_block_1/use_batchnorm': False,
     'text_block_1/dense_block_1/dropout_rate': 0.5,
     'text_block_1/dense_block_1/units_0': 256,
+}, {
```
Comment (haifeng-jin) [inline autokeras/tuners/task_specific.py:53]: We need a test for this part, too. Please refer to task_specific_test.py

File: autokeras/blocks/basic.py [hunk 2]
```diff
@@ -375,6 +553,86 @@ def build(self, hp, inputs=None):
...
+            'embedding_dim',
+            [32, 64, 128, 256, 512],
+            default=128)
+
+        if self.dropout_rate is not None:
+            dropout_rate = self.dropout_rate
+        else:
+            dropout_rate = hp.Choice('dropout_rate', [0.0, 0.25, 0.5], default=0.25)
+
+        TokenEmbedding = Embedding(max_features=self.max_features,
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:621]: For variables, we use `token_embedding` and `position_embedding`.

File: autokeras/blocks/basic.py [hunk 3]
```diff
@@ -375,6 +553,86 @@ def build(self, hp, inputs=None):
...
+        # Arguments
+             hp: HyperParameters. The hyperparameters for building the model.
+             inputs: Tensor of Shape [batch_size, seq_len, embedding_dim]
+
+        # Returns
+            Output Tensor of shape `[batch_size, seq_len, embedding_dim]`.
+        """
+        input_node = nest.flatten(inputs)[0]
+        # TODO: support more pretrained embedding layers.
+        # glove, fasttext, and word2vec
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:606]: Remove these 2 lines.

File: autokeras/blocks/basic.py [hunk 4]
```diff
@@ -375,6 +553,86 @@ def build(self, hp, inputs=None):
         return model.outputs
 
 
+class TokenAndPositionEmbedding(block_module.Block):
+    """
+    Token and Position Embedding block for Transformer.
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:558]: Make them one line.

File: autokeras/blocks/basic.py [hunk 5]
```diff
@@ -260,6 +261,183 @@ def _get_padding(kernel_size, output_node):
...
+
+    @staticmethod
+    def separate_heads(x, batch_size, num_heads, projection_dim):
+        x = tf.reshape(x, (batch_size, -1, num_heads, projection_dim))
+        return tf.transpose(x, perm=[0, 2, 1, 3])
+
+
+class TransformerBlock(block_module.Block):
+    """
+        Block for Transformer.
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:364]: Make them one line.

File: autokeras/blocks/__init__.py [hunk 6]
```diff
@@ -3,8 +3,10 @@
 from autokeras.blocks.basic import ConvBlock
 from autokeras.blocks.basic import DenseBlock
 from autokeras.blocks.basic import Embedding
+from autokeras.blocks.basic import MultiHeadSelfAttentionBlock
```
Comment (haifeng-jin) [inline autokeras/blocks/__init__.py:6]: Let's not expose it for now.

**Comment by achaar** [comment]
The performance is almost similar to the tutorial now.

File: autokeras/blocks/wrapper.py [hunk 7]
```diff
@@ -100,6 +105,9 @@ def get_config(self):
     def build(self, hp, inputs=None):
         input_node = nest.flatten(inputs)[0]
         output_node = input_node
+        block_type = self.block_type or hp.Choice('block_type',
+                                                  ['vanilla', 'transformer'],
+                                                  default='vanilla')
         vectorizer = self.vectorizer or hp.Choice('vectorizer',
                                                   ['sequence', 'ngram'],
                                                   default='sequence')
```
Comment (haifeng-jin) [inline autokeras/blocks/wrapper.py:113]: Make these two one hp with name block_type with 3 values.

File: autokeras/blocks/basic.py [hunk 8]
```diff
@@ -260,6 +261,181 @@ def _get_padding(kernel_size, output_node):
...
+    def separate_heads(x, batch_size, num_heads, projection_dim):
+        x = tf.reshape(x, (batch_size, -1, num_heads, projection_dim))
+        return tf.transpose(x, perm=[0, 2, 1, 3])
+
+
+class TransformerBlock(block_module.Block):
+    """Block for Transformer.
+
+    # Arguments
+    embed_dim: Int. Output dimension of the Attention block.
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:364]: add indent

File: autokeras/blocks/basic.py [hunk 9]
```diff
@@ -375,6 +551,93 @@ def build(self, hp, inputs=None):
         return model.outputs
 
 
+class TokenAndPositionEmbedding(block_module.Block):
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:554]: I still feel we should make it part of the Transformer block, unless it is useful by any other well know models.

File: autokeras/blocks/basic.py [hunk 10]
```diff
@@ -260,6 +261,221 @@ def _get_padding(kernel_size, output_node):
...
+    @staticmethod
+    def separate_heads(x, batch_size, num_heads, projection_dim):
+        x = tf.reshape(x, (batch_size, -1, num_heads, projection_dim))
+        return tf.transpose(x, perm=[0, 2, 1, 3])
+
+
+class TransformerBlock(block_module.Block):
+    """Block for Transformer.
+    The input should be tokenized sequences with the same length, where each element
+    of a sequence should be the index of the word.
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:363]: Let's add an example here of using AutoModel to connect Transformer. Please refer to `AutoModel`'s docstring.

File: autokeras/blocks/basic.py [hunk 11]
```diff
@@ -260,6 +261,221 @@ def _get_padding(kernel_size, output_node):
...
+        output = tf.matmul(weights, value)
+        return output, weights
+
+    @staticmethod
+    def separate_heads(x, batch_size, num_heads, projection_dim):
+        x = tf.reshape(x, (batch_size, -1, num_heads, projection_dim))
+        return tf.transpose(x, perm=[0, 2, 1, 3])
+
+
+class TransformerBlock(block_module.Block):
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:360]: Also please rename `TransformerBlock` to `Transformer`. We usually only add a "Block" as suffix whenever the name is confusing without it.

File: autokeras/blocks/basic.py [hunk 12]
```diff
@@ -260,6 +261,221 @@ def _get_padding(kernel_size, output_node):
         return 'same'
 
 
+class MultiHeadSelfAttentionBlock(block_module.Block):
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:264]: Rename it to MultiHeadSelfAttention.

File: autokeras/blocks/basic.py [hunk 13]
```diff
@@ -260,6 +261,222 @@ def _get_padding(kernel_size, output_node):
         return 'same'
 
 
+class MultiHeadSelfAttention(block_module.Block):
```
Comment (fchollet) [inline autokeras/blocks/basic.py:264]: Please make sure to follow the API for the official MultiHeadAttention layer. https://github.com/tensorflow/community/pull/260

File: autokeras/blocks/basic.py [hunk 14]
```diff
@@ -260,6 +261,222 @@ def _get_padding(kernel_size, output_node):
...
+        max_features: Int. Size of the vocabulary. Must be set if not using
+            TextToIntSequence before this block. Defaults to 20001.
+        pretraining: String. 'random' (use random weights instead any pretrained
+            model), 'glove', 'fasttext' or 'word2vec'. Use pretrained word embedding.
+            If left unspecified, it will be tuned automatically.
+        embedding_dim: Int. Output dimension of the Attention block.
+            If left unspecified, it will be tuned automatically.
+        num_heads: Int. The number of attention heads. If left unspecified,
+            it will be tuned automatically.
+        ff_dim: Int. The output dimension of the FFN. If left
```
Comment (fchollet) [inline autokeras/blocks/basic.py:375]: Always use fully-spelled out argument names instead of abbreviations, unless the abbreviation is already known to 100% of people who will read it.

File: autokeras/blocks/basic.py [hunk 15]
```diff
@@ -260,6 +261,222 @@ def _get_padding(kernel_size, output_node):
...
+            unspecified, it will be tuned automatically.
+        dropout_rate: Float. Between 0 and 1. If left unspecified, it will be
```
Comment (fchollet) [inline autokeras/blocks/basic.py:377]: This is usually named `dropout` -- let's be consistent with the core API.

File: autokeras/blocks/basic.py (hunk 13 above)
Comment (haifeng-jin) [reply autokeras/blocks/basic.py:264]: Currently, `MultiHeadSelfAttention` is not exposed to the users. The `MultiHeadAttention` is more generic since it is not just self attention. We can try to make them similar without making it generic by only adding the args we used.

File: autokeras/blocks/basic.py (hunk 15 above)
Comment (haifeng-jin) [reply autokeras/blocks/basic.py:396]: This will be in another PR, since there are too many renamings in other parts of the code.

File: autokeras/blocks/basic.py [hunk 16]
```diff
@@ -260,6 +261,243 @@ def _get_padding(kernel_size, output_node):
...
+            default=128)
+        num_heads = self.num_heads
+        if num_heads is None:
+            num_heads = 8
+
+        if head_size % num_heads != 0:  # how to evaluate this condition
+            raise ValueError(
+                f"embedding dimension = {head_size} should be "
+                f"divisible by number of heads = {num_heads}"
+            )
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:321]: Please change it to the same format as the ValueError on line 304. single quote, and .format

Comment (haifeng-jin) [inline autokeras/blocks/basic.py:321]: Please make this  the same format as the one on 304. single quote, use .format. Also please test all the error messages in this PR in the unit tests.

File: autokeras/blocks/basic.py [hunk 17]
```diff
@@ -260,6 +261,243 @@ def _get_padding(kernel_size, output_node):
...
+        super().__init__(**kwargs)
+        self.head_size = head_size
+        self.num_heads = num_heads
+
+    def get_config(self):
+        config = super().get_config()
+        config.update({
+            'head_size': self.head_size,
+            'num_heads': self.num_heads})
+        return config
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:288]: These lines needs to be covered by the tests.
